mpv.py

```diff
@@ -1040,6 +1054,20 @@ def command(self, name, *args):
                 for arg in args if arg is not None ] + [None]
         _mpv_command(self.handle, (c_char_p*len(args))(*args))
 
+    def command_async(self, name, *args, callback=None):
+        """
+        Same as mpv_command, but run the command asynchronously.
+        Once the command ran, the callback will be invoked, if provided.
+        The first argument of the callback will be a boolean value.
+        It will be set to True, if there was an error, False else.
+        The second argument of the callback depends on the command.
+        """
+        args = [name.encode('utf-8')] + [(arg if type(arg) is bytes else str(arg).encode('utf-8'))
+                                         for arg in args if arg is not None] + [None]
+        key = randint(0, 2147483647)
```

Comment:
I think using hash(callback) here would be better. That value is guaranteed to be unique, so even something calling millions of async commands will never see a collision. Also, the lifetime of a hash(obj) return value happens to align with the lifetime of the object, in this case the callback. Thus, hash values will naturally be recycled when the callback is deleted after it is removed from the callback registry after it has been called. mpv's client.h specifies that userdata is indeed a 64-bit value.

Reply:
I agree that `randint(0, 2147483647)` is not optimal here. But if we use the hash of the callback we might accidentally drop some callbacks.

Consider the function call:
```py
self.mpv.command_async("keydown", "MOUSE_BTN" + str(0), callback=self._button_press_cb)
```
The hash of the callback function always stays the same, and we can not guarantee, that our event thread calls back fast enough, so that the key (the hash of the callback that is) is 'unused' (removed from the dict with callbacks) again. I was able to provoke the following output for fast consecutive events (mouse move events):
```
pop 8793910595232
Put 8793910595232
Put 8793910595232
pop 8793910595232
pop 8793910595232
```

So maybe we could use some sort of atomic counter or lock?

```python
with some_lock:
    async_counter = (async_counter + 1) % 2147483647
```

Reply:
Hmm, true. The counter idea is good, but it should wrap at 2**64, not 2**31-1 since command_async's userdata is an uint64_t. An alternative would be to wrap every callback in a new lambda, but I think even with the lock your solution should be faster.

Reply:
Okay! I don't know what would be fast here: lambda or lock.

What do you think of the following solution:
```python
    def _register_async_callback(self, callback):
        if callback is None:
            key = 18_446_744_073_709_551_616
        else:
            with self._event_async_callback_counter_lock:
                key = self._event_async_callback_counter \
                    = (self._event_async_callback_counter + 1) % 18_446_744_073_709_551_615
        self._event_async_callbacks[key] = callback
        return key
```
If we reserve the highest uint64_t value for the case where we don't have a callback we can avoid the lock and since we don't need to call back, we don't care if we override it or not :sweat_smile: 

As there probably aren't many commands that actually return a value we could also use the lambda solution.

Reply:
I've force pushed the changes (lock solution). Let me know if you're happy with it, or if you want me to change something :smiley:

Reply:
Okay, in the last push, I removed the special treatment of the None callback and factored out the constant (we only need one :see_no_evil:).